### PR TITLE
Avoid %% escape in xtrigger func args (7.8.x).

### LINKED ISSUE #3256: Allow proper escaping in xtrigger format statements
I had a format string that I wanted to pass through the xtrigger formatting, but the regular expression used to match format strings and check their validity did not honor the Python escape sequence, i.e.
"%%(asdf)s" after running through string format is "%(asdf)s", but the current xtrigger format check in the suite validation complains that "asdf" isn't one of the approved variables for string substitution.

Use case here is to allow for parameterized filenames in the xtrigger, but I can't directly use cycle points as the files have 10-digit YYYYMMDDHH in the name so I resorted to passing in a format-string filename and a date-time-group format.

### PR DESCRIPTION
**Pull request by hjoliver**
These changes close #3256

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).

- [x] Appropriate tests are included (unit and/or functional).

- [x] Includes an appropriate entry in the release change log `CHANGES.md`.

- [x] (7.8.x branch) I have updated the documentation in this PR branch.

### CODE REVIEW

**Comment by hjoliver** [comment]
Test using the built-in `echo` xtrigger, which just echoes its function args (run in debug mode to see the result in the server log):
```
# workflow "test3256"

[scheduling]
  [[xtriggers]]
     a = echo(%(point)s, '%%(asdf)s')
  [[dependencies]]
     graph = "@a => foo"
[runtime]
  [[foo]]
     script = sleep 10
```

In the function args, `%(point)s` is a valid string template for cycle point; `%%(asdf)s` is not a template because `%` is escaped, so it should be passed in as a string literal.

On current 7.8.x and master branches (reproduces the bug):
```
[oliverh@niwa-1007885 cylc-flow]$ cylc val test3256
Illegal template in xtrigger a: asdf
```

On this branch, validation and run:
```
[oliverh@niwa-1007885 cylc-flow]$ cylc val test3256
Valid for cylc-7.8.2-32-g32f85

[oliverh@niwa-1007885 cylc-flow]$ cylc run --no-detach --debug test3256
2019-07-30T13:51:23+12:00 DEBUG - Loading site/user global config files
2019-07-30T13:51:23+12:00 DEBUG - Loading site/user global config files
            ._.                                                       
            | |            The Cylc Suite Engine [7.8.2-32-g32f85]    
._____._. ._| |_____.           Copyright (C) 2008-2019 NIWA          
| .___| | | | | .___|   & British Crown (Met Office) & Contributors.  
| !___| !_! | | !___.  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
!_____!___. |_!_____!  This program comes with ABSOLUTELY NO WARRANTY;
      .___! |          see `cylc warranty`.  It is free software, you 
      !_____!           are welcome to redistribute it under certain  
2019-07-30T13:51:23+12:00 DEBUG - creating suite run directory: /home/oliverh/cylc-run/test3256
...(SNIP)...
2019-07-30T13:51:24+12:00 DEBUG - [xtrigger-func cmd] cylc-function-run echo '["1", "%(asdf)s"]' '{}' /home/oliverh/suites/foo
        [xtrigger-func ret_code] 0
        [xtrigger-func out] [false, {}]
        [xtrigger-func err]
        echo: ARGS: (u'1', u'%(asdf)s')  <-----------------------BINGO!
        echo: KWARGS: {}
2019-07-30T13:51:24+12:00 DEBUG - echo(1, %(asdf)s): returned {}
```

File: lib/cylc/xtrigger_mgr.py [hunk 1]
```diff
@@ -40,8 +40,9 @@
     TMPL_TASK_CYCLE_POINT, TMPL_TASK_IDENT, TMPL_TASK_NAME, TMPL_SUITE_RUN_DIR,
     TMPL_SUITE_SHARE_DIR, TMPL_USER_NAME, TMPL_SUITE_NAME, TMPL_DEBUG_MODE]
 
-# Extract all 'foo' from string templates '%(foo)s'.
-RE_STR_TMPL = re.compile(r'%\(([\w]+)\)s')
+# Extract 'foo' from string templates '%(foo)s', avoiding '%%' escaping
+# ('%%(foo)s` is not a string template).
+RE_STR_TMPL = re.compile(r'(?<!%)%\(([\w]+)\)s')
```
Comment (hjoliver) [inline lib/cylc/xtrigger_mgr.py]: Zero-width negative lookahead assertion, innit :grin:

**Comment by trwhitcomb** [comment]
You beat me to it!  I used an exclusion class (i.e. `r[^%]%\(`) instead of any fancy lookbehinds though.

**Comment by hjoliver** [comment]
> You beat me to it! I used an exclusion class (i.e. `r[^%]%\(`) instead of any fancy lookbehinds though.

Sorry about that!  However, I don't think the exclusion class works because it is not zero width. i.e. it would match `x%(blah)s` but not plain old `%(blah)s`.  (And adding a "zero or one of" qualifier doesn't fix the problem because the other % gets in the way then?)

**Comment by trwhitcomb** [comment]
I think you're right, but because the exclusion isn't part of the capturing group, even if it did match `x%(blah)s` it should still only pick up "blah" which is what tjis code is looking for. 

I don't have an issue with the zero width lookaround, but I will confess that I didn't recognize it immediately.

**Comment by kinow** [comment]
Maybe we should include a few unit tests covering the tests above :point_up: ? :grimacing:

**Comment by hjoliver** [comment]
> Maybe we should include a few unit tests covering the tests above point_up ? grimacing

Yes, I've left this as a WIP/Draft pending time to think about tests.

**Comment by kinow** [comment]
Oh, sorry, didn't look at the checklist.

**Comment by trwhitcomb** [comment]
Now I'm very glad my fix got scooped by @hjoliver!

**Comment by hjoliver** [comment]
God damn it.  Style fail.

**Comment by hjoliver** [comment]
@trwhitcomb - I've added you as a "collaborator" on this repository so I can assign you to review this PR. You should get an invite from GitHub.

**Comment by matthewrmshin** [comment]
Test failures are related to #3244 - Travis CI hostname issue.

**Pull request merged**